File: src/2020/day1/report_repair.py
import functools
import itertools
from operator import mul
from pathlib import Path


def report_repair(expense_report: Path, combinations_length: int) -> None:
    with open(expense_report) as file:
        expense_report_vals = [int(line) for line in file.read().splitlines()]
    for vals in itertools.combinations(expense_report_vals, combinations_length):
        if sum(vals) == 2020:
            print(
                f"Values {vals} sum to 2020 and multiplied give {functools.reduce(mul, vals)}"
            )


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="AoC 2020 - Day 1: Report Repair")
    parser.add_argument(
        "report_path", type=Path, help="The path to the report file to read"
    )
    parser.add_argument(
        "combinations_length",
        type=int,
        help="The length of combinations of expenses to consider",
    )
    args = parser.parse_args()
    report_repair(args.report_path, args.combinations_length)

File: src/2020/day1/report_repair.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use itertools::Itertools;
use structopt::StructOpt;

fn report_repair(expense_report: &Path, combinations_length: u8) {
    let expense_report_vals: Vec<usize> = fs::read_to_string(expense_report)
        .expect(format!("No file found at: {:?}.", expense_report).as_ref())
        .lines()
        .map(|x| {
            usize::from_str(x).expect("Report expense entry couldn't be converted to `usize`.")
        })
        .collect();
    for vals in expense_report_vals
        .into_iter()
        .combinations(combinations_length as usize)
    {
        let sum: usize = vals.iter().sum();
        if sum == 2020 {
            let product: usize = vals.iter().product();
            println!(
                "Values {:?} sum to 2020 and multiplied give {}",
                vals, product
            )
        }
    }
}

/// AoC 2020 - Day 1: Report Repair
#[derive(StructOpt)]
struct Cli {
    /// The path to the report file to read
    #[structopt(parse(from_os_str))]
    report_path: PathBuf,
    /// The length of combinations of expenses to consider
    combinations_length: u8,
}

fn main() {
    let args = Cli::from_args();
    report_repair(&args.report_path, args.combinations_length);
}
